File: snake_game.js
// snake_game.js

const canvas = document.getElementById('gameCanvas');
const ctx = canvas.getContext('2d');

const GRID_SIZE = 20;
const GRID_WIDTH = canvas.width / GRID_SIZE;
const GRID_HEIGHT = canvas.height / GRID_SIZE;
const FPS = 10;

let snake = {
    length: 1,
    positions: [{ x: Math.floor(GRID_WIDTH / 2), y: Math.floor(GRID_HEIGHT / 2) }],
    direction: { x: 0, y: 0 },
    color: 'green',
    score: 0,
    foodPosition: { x: 0, y: 0 },
};

function drawSnake() {
    ctx.fillStyle = snake.color;
    ctx.strokeStyle = 'white';
    snake.positions.forEach(pos => {
        ctx.fillRect(pos.x * GRID_SIZE, pos.y * GRID_SIZE, GRID_SIZE, GRID_SIZE);
        ctx.strokeRect(pos.x * GRID_SIZE, pos.y * GRID_SIZE, GRID_SIZE, GRID_SIZE);
    });
}

function drawFood() {
    ctx.fillStyle = 'red';
    ctx.fillRect(snake.foodPosition.x * GRID_SIZE, snake.foodPosition.y * GRID_SIZE, GRID_SIZE, GRID_SIZE);
}

function drawScore() {
    ctx.fillStyle = 'white';
    ctx.font = '20px Arial';
    ctx.fillText(`Score: ${snake.score}`, 10, 30);
}

function drawTitle() {
    ctx.fillStyle = 'white';
    ctx.font = '30px Arial';
    ctx.fillText('Snake Game', 200, 30);
}

function handleKeys(event) {
    switch (event.key) {
        case 'ArrowUp':
            snake.direction = { x: 0, y: -1 };
            break;
        case 'ArrowDown':
            snake.direction = { x: 0, y: 1 };
            break;
        case 'ArrowLeft':
            snake.direction = { x: -1, y: 0 };
            break;
        case 'ArrowRight':
            snake.direction = { x: 1, y: 0 };
            break;
    }
}

function moveSnake() {
    const head = { x: snake.positions[0].x + snake.direction.x, y: snake.positions[0].y + snake.direction.y };
    snake.positions.unshift(head);

    if (head.x === snake.foodPosition.x && head.y === snake.foodPosition.y) {
        snake.score++;
        snake.length++;
        generateFoodPosition();
    } else {
        snake.positions.pop();
    }

    if (isSnakeCollided()) {
        resetGame();
    }
}

function isSnakeCollided() {
    const head = snake.positions[0];
    return (
        head.x < 0 ||
        head.x >= GRID_WIDTH ||
        head.y < 0 ||
        head.y >= GRID_HEIGHT ||
        snake.positions.slice(1).some(segment => segment.x === head.x && segment.y === head.y)
    );
}

function generateFoodPosition() {
    snake.foodPosition = {
        x: Math.floor(Math.random() * GRID_WIDTH),
        y: Math.floor(Math.random() * GRID_HEIGHT),
    };

    if (snake.positions.some(pos => pos.x === snake.foodPosition.x && pos.y === snake.foodPosition.y)) {
        generateFoodPosition();
    }
}

function resetGame() {
    snake = {
        length: 1,
        positions: [{ x: Math.floor(GRID_WIDTH / 2), y: Math.floor(GRID_HEIGHT / 2) }],
        direction: { x: 0, y: 0 },
        color: 'green',
        score: 0,
        foodPosition: { x: 0, y: 0 },
    };
}

function draw() {
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    drawTitle();
    drawScore();
    drawSnake();
    drawFood();
}

function main() {
    document.addEventListener('keydown', handleKeys);
    setInterval(() => {
        moveSnake();
        draw();
    }, 1000 / FPS);
}

main();

File: snake_game.py
# snake_game.py

import pygame
import sys
import secrets

# Initialize Pygame
pygame.init()

# Constants
SCREEN_WIDTH = 600
SCREEN_HEIGHT = 400
GRID_SIZE = 20
GRID_WIDTH = SCREEN_WIDTH // GRID_SIZE
GRID_HEIGHT = SCREEN_HEIGHT // GRID_SIZE
FPS = 10

# Colors
BLACK = (0, 0, 0)
WHITE = (255, 255, 255)
RED = (255, 0, 0)
GREEN = (0, 255, 0)

# Snake class
class Snake:
    def __init__(self):
        self.length = 1
        self.positions = [((SCREEN_WIDTH // 2), (SCREEN_HEIGHT // 2))]
        self.direction = secrets.choice([pygame.K_UP, pygame.K_DOWN, pygame.K_LEFT, pygame.K_RIGHT])
        self.color = GREEN
        self.score = 0
        self.food_position = self.random_food_position()

    def get_head_position(self):
        return self.positions[0]

    def turn(self, point):
        if self.length > 1 and (point[0]*-1, point[1]*-1) == self.direction:
            return
        else:
            self.direction = point

    def move(self):
        cur = self.get_head_position()
        x, y = self.direction
        new = (((cur[0] + (x*GRID_SIZE)) % SCREEN_WIDTH), (cur[1] + (y*GRID_SIZE)) % SCREEN_HEIGHT)
        if len(self.positions) > 2 and new in self.positions[2:]:
            self.reset()
        else:
            self.positions.insert(0, new)
            if len(self.positions) > self.length:
                self.positions.pop()

    def reset(self):
        self.length = 1
        self.positions = [((SCREEN_WIDTH // 2), (SCREEN_HEIGHT // 2))]
        self.direction = secrets.choice([pygame.K_UP, pygame.K_DOWN, pygame.K_LEFT, pygame.K_RIGHT])
        self.score = 0
        self.food_position = self.random_food_position()

    def random_food_position(self):
        return (secrets.SystemRandom().randint(0, GRID_WIDTH-1)*GRID_SIZE, secrets.SystemRandom().randint(0, GRID_HEIGHT-1)*GRID_SIZE)

    def draw(self, surface):
        for p in self.positions:
            r = pygame.Rect((p[0], p[1]), (GRID_SIZE, GRID_SIZE))
            pygame.draw.rect(surface, self.color, r)
            pygame.draw.rect(surface, WHITE, r, 1)

    def handle_keys(self):
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                sys.exit()
            elif event.type == pygame.KEYDOWN:
                if event.key == pygame.K_q:
                    pygame.quit()
                    sys.exit()
                elif event.key == pygame.K_LEFT:
                    self.turn((-1, 0))
                elif event.key == pygame.K_RIGHT:
                    self.turn((1, 0))
                elif event.key == pygame.K_UP:
                    self.turn((0, -1))
                elif event.key == pygame.K_DOWN:
                    self.turn((0, 1))

    def check_food(self):
        if self.get_head_position() == self.food_position:
            self.score += 1
            self.length += 1
            self.food_position = self.random_food_position()

    def draw_food(self, surface):
        r = pygame.Rect((self.food_position[0], self.food_position[1]), (GRID_SIZE, GRID_SIZE))
        pygame.draw.rect(surface, RED, r)
        pygame.draw.rect(surface, WHITE, r, 1)

# Main function to run the game
def main():
    clock = pygame.time.Clock()
    screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
    pygame.display.set_caption('Snake Game')

    snake = Snake()

    while True:
        snake.handle_keys()
        snake.move()
        snake.check_food()

        screen.fill(BLACK)
        snake.draw(screen)
        snake.draw_food(screen)

        pygame.display.update()
        clock.tick(FPS)

if __name__ == '__main__':
    main()
